bench/Bench/Utils.hs

```diff
@@ -0,0 +1,176 @@
+
+{-------------------------------------------------------------------------------
+  Environment types
+-------------------------------------------------------------------------------}
+
+data BenchEnv k v = BenchEnv {
+    dbFilePath :: FilePath
+  , dbEnv      :: Environment ReadWrite
+  , db         :: Database k v
+  }
+
+-- TODO: is this instance sufficient? I think so, since a @'BenchEnv'@ mainly
+-- consists of pointers.
+instance NFData (BenchEnv k v) where
+  rnf :: BenchEnv k v -> ()
+  rnf _ = ()
+
+{-------------------------------------------------------------------------------
+  Initialisation and cleanup for environments
+-------------------------------------------------------------------------------}
+
+initBenchEnv ::
+     (MonadIO m, Serialise k, Serialise v, Arbitrary k, Arbitrary v)
+  => Proxy k
+  -> Proxy v
+  -> Limits
+     -- ^ LMDB limits such as maximum database size.
+  -> (Database k v -> Transaction ReadWrite ())
+     -- ^ Action that populate the database
+  -> m (BenchEnv k v)
+initBenchEnv _ _ limits populate = do
+  dbFilePath <- createDbDir
+  dbEnv <- liftIO $ openEnvironment dbFilePath limits
+  db <- liftIO $ readWriteTransaction dbEnv (getDatabase Nothing)
+  void $ liftIO $ readWriteTransaction dbEnv $ populate db
+  pure BenchEnv {dbFilePath, dbEnv, db}
+
+createDbDir :: MonadIO m => m FilePath
+createDbDir = do
+  sysTmpDir <- liftIO getTemporaryDirectory
+  liftIO $ createTempDirectory sysTmpDir "bench_cursor"
+
+cleanupBenchEnv :: MonadIO m => BenchEnv k v -> m ()
+cleanupBenchEnv tenv = do
+  liftIO $ closeEnvironmentBlocking (dbEnv tenv)
+  deleteDbDir (dbFilePath tenv)
+
+deleteDbDir :: MonadIO m => FilePath -> m ()
+deleteDbDir = liftIO . removeDirectoryRecursive
+
+{-------------------------------------------------------------------------------
+  Defaults for LMDB limits
+-------------------------------------------------------------------------------}
+
+simpleLMDBLimits :: Limits
+simpleLMDBLimits = defaultLimits
+
+{-------------------------------------------------------------------------------
+  Defaults for database population
+-------------------------------------------------------------------------------}
+
+-- | @'noPopulateDb' db@ does not populate the LMDB database @db@ with any
+-- entries.
+noPopulateDb :: Database k v -> Transaction ReadWrite ()
+noPopulateDb _ = pure ()
+
+-- | @'populateDbUnique' n db@ populates the LMDB database @db@ with @n@ unique
+-- key-value pairs.
+--
+-- Note: ensure that @n@ is much smaller than the range of values that are
+-- generated for @'Arbitrary' k@, or this function may not terminate within
+-- reasonable time, let alone terminate at all. We use a simple "retry if a
+-- generated key is already present in the database (i.e., a /clash/)" strategy
+-- to generate exactly @n@ unique keys, which means that clashes occur
+-- inhibitively often as @n@ approaches the size of the of the range of @k@.
+populateDbUnique ::
```

Comment:
Maybe we should have an assertion stating that `n <= 2^32`. As for the termination problem, what about keeping track of the lowest value not in the DB (initially `0x0`) and if we find a clash go with the next value?

Reply:
An assertion would make sense, but `populateDbUnique` is polymorphic in the key type @k@ so we can't make assumptions about the upper bound on the number of unique keys

Reply:
Regarding your idea for the termination problem, that would improve termination guarantees, but it would probably skew the distribution of keys towards smaller keys (since we always pick a small key in case of a clash). I'm going to leave the function as is for now